File: netflix_analysis.R
pdf(NULL)


library(ggplot2)
library(dplyr)
library(readr)
library(viridis)

cat("Starting R analysis of Netflix data...\n")

if (file.exists("data/Netflix_shows_movies.csv")) {
  netflix_data <- read_csv("data/Netflix_shows_movies.csv")
  cat("Loaded original dataset - will do basic cleaning\n")
  
  # Quick cleaning
  netflix_data$director[is.na(netflix_data$director)] <- "Unknown"
  netflix_data$cast[is.na(netflix_data$cast)] <- "Unknown"
  netflix_data$country[is.na(netflix_data$country)] <- "Unknown"
  netflix_data$rating[is.na(netflix_data$rating)] <- "Not Rated"
  netflix_data$listed_in[is.na(netflix_data$listed_in)] <- "Unknown"
} else {
  stop("Cannot find dataset file!")
}

cat(paste("Dataset has", nrow(netflix_data), "rows and", ncol(netflix_data), "columns\n"))

cat("Creating ratings visualization...\n")

ratings_summary <- netflix_data %>%
  count(rating, sort = TRUE) %>%
  mutate(percentage = n / sum(n) * 100)

# Create ggplot visualization
ratings_plot <- ggplot(ratings_summary, aes(x = reorder(rating, n), y = n, fill = rating)) +
  geom_col(width = 0.8, alpha = 0.8) +
  geom_text(aes(label = paste(n, "\n(", round(percentage, 1), "%)", sep = "")), 
            hjust = -0.1, size = 3) +
  coord_flip() +
  labs(title = "Netflix Content Rating Distribution",
       subtitle = "Analysis using R and ggplot2",
       x = "Content Rating",
       y = "Number of Titles") +
  theme_minimal() +
  theme(plot.title = element_text(size = 14, face = "bold"),
        plot.subtitle = element_text(size = 11),
        axis.title = element_text(size = 11),
        legend.position = "none") +
  scale_fill_viridis_d()

print(ratings_plot)

ggsave("charts/netflix_ratings_R_version.png", ratings_plot, width = 10, height = 6, dpi = 300)
cat("Ratings plot saved as charts/netflix_ratings_R_version.png\n")

cat("Creating genre analysis...\n")

all_genres <- c()
for (i in 1:nrow(netflix_data)) {
  if (!is.na(netflix_data$listed_in[i])) {
    genres <- strsplit(as.character(netflix_data$listed_in[i]), ",")[[1]]
    genres <- trimws(genres)  # Remove whitespace
    all_genres <- c(all_genres, genres)
  }
}

# Count and get top genres
genre_counts <- table(all_genres)
genre_counts <- sort(genre_counts, decreasing = TRUE)
top_genres <- head(genre_counts, 12)  # Top 12 for better display

# Create data frame for plotting
genres_expanded <- data.frame(
  listed_in = names(top_genres),
  n = as.numeric(top_genres)
) %>%
  mutate(listed_in = reorder(listed_in, n))

# Create genre plot
genre_plot <- ggplot(genres_expanded, aes(x = listed_in, y = n)) +
  geom_col(fill = "steelblue", alpha = 0.7) +
  geom_text(aes(label = n), hjust = -0.2, size = 3) +
  coord_flip() +
  labs(title = "Top Netflix Genres",
       x = "Genre",
       y = "Number of Titles") +
  theme_minimal() +
  theme(plot.title = element_text(size = 14, face = "bold"))

print(genre_plot)
ggsave("charts/netflix_genres_R_version.png", genre_plot, width = 10, height = 6, dpi = 300)
cat("Genre plot saved as charts/netflix_genres_R_version.png\n")

cat("\n--- R ANALYSIS SUMMARY ---\n")
cat(paste("Total titles analyzed:", nrow(netflix_data), "\n"))

movie_count <- sum(netflix_data$type == "Movie", na.rm = TRUE)
tv_count <- sum(netflix_data$type == "TV Show", na.rm = TRUE)

cat(paste("Movies:", movie_count, "(", round(movie_count/nrow(netflix_data)*100, 1), "%)\n"))
cat(paste("TV Shows:", tv_count, "(", round(tv_count/nrow(netflix_data)*100, 1), "%)\n"))

cat("\nTop 5 ratings:\n")
top_ratings <- head(ratings_summary, 5)
for(i in 1:nrow(top_ratings)) {
  cat(paste(i, ".", top_ratings$rating[i], ":", top_ratings$n[i], "titles\n"))
}

cat("\nFiles created:\n")
cat("- charts/netflix_ratings_R_version.png\n")
cat("- charts/netflix_genres_R_version.png\n")

cat("\nR analysis completed!\n")

File: netflix_analysis.py
import matplotlib
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter
import warnings
warnings.filterwarnings('ignore')

plt.style.use('default')
sns.set_palette("husl")
matplotlib.use('Agg')  # Prevents plot windows from opening
plt.ioff()  # Turn off interactive mode

print("Starting Netflix Data Analysis...")
print("=" * 40)

def load_data(filename):
    try:
        df = pd.read_csv(filename)
        print(f"Successfully loaded dataset: {df.shape[0]} rows, {df.shape[1]} columns")
        return df
    except FileNotFoundError:
        print("Error: Could not find the dataset file!")
        return None

Netflix_shows_movies = load_data('data/Netflix_shows_movies.csv')

if Netflix_shows_movies is not None:
    # Quick look at the data
    print("\nFirst few rows:")
    print(Netflix_shows_movies.head())
    
    print("\nDataset info:")
    print(Netflix_shows_movies.info())

# Handle missing values
print("\n" + "=" * 40)
print("DATA CLEANING SECTION")
print("=" * 40)

print("Missing values before cleaning:")
print(Netflix_shows_movies.isnull().sum())

# Fill missing values - different approach for each column
Netflix_shows_movies['director'].fillna('No Director Listed', inplace=True)
Netflix_shows_movies['cast'].fillna('No Cast Info', inplace=True)
Netflix_shows_movies['country'].fillna('Country Not Specified', inplace=True)
Netflix_shows_movies['date_added'].fillna('Date Unknown', inplace=True)
Netflix_shows_movies['rating'].fillna('Not Rated', inplace=True)

# For duration - fill based on content type
movie_duration = Netflix_shows_movies[Netflix_shows_movies['type'] == 'Movie']['duration'].mode()
if len(movie_duration) > 0:
    Netflix_shows_movies.loc[(Netflix_shows_movies['type'] == 'Movie') & 
                           (Netflix_shows_movies['duration'].isnull()), 'duration'] = movie_duration[0]

tv_duration = Netflix_shows_movies[Netflix_shows_movies['type'] == 'TV Show']['duration'].mode()
if len(tv_duration) > 0:
    Netflix_shows_movies.loc[(Netflix_shows_movies['type'] == 'TV Show') & 
                           (Netflix_shows_movies['duration'].isnull()), 'duration'] = tv_duration[0]

Netflix_shows_movies['duration'].fillna('Duration Unknown', inplace=True)
Netflix_shows_movies['listed_in'].fillna('Genre Not Listed', inplace=True)
Netflix_shows_movies['description'].fillna('No description available', inplace=True)

print("\nMissing values after cleaning:")
print(Netflix_shows_movies.isnull().sum())

print("\n" + "=" * 40)
print("DATA EXPLORATION")
print("=" * 40)

print("Basic dataset statistics:")
print(f"Total number of titles: {len(Netflix_shows_movies)}")

# Content type breakdown
content_types = Netflix_shows_movies['type'].value_counts()
print(f"\nContent breakdown:")
for content_type, count in content_types.items():
    percentage = (count / len(Netflix_shows_movies)) * 100
    print(f"{content_type}: {count} ({percentage:.1f}%)")

# Rating distribution
print(f"\nTop 10 ratings:")
rating_counts = Netflix_shows_movies['rating'].value_counts().head(10)
print(rating_counts)

# Release year stats
print(f"\nRelease year information:")
print(f"Oldest content: {Netflix_shows_movies['release_year'].min()}")
print(f"Newest content: {Netflix_shows_movies['release_year'].max()}")
print(f"Average release year: {Netflix_shows_movies['release_year'].mean():.1f}")

# Country analysis
print(f"\nTop countries (handling multiple countries):")
all_countries = []
for country_list in Netflix_shows_movies['country'].dropna():
    if ',' in str(country_list):
        countries = [c.strip() for c in str(country_list).split(',')]
        all_countries.extend(countries)
    else:
        all_countries.append(str(country_list).strip())

country_counter = Counter(all_countries)
top_countries = country_counter.most_common(10)
for country, count in top_countries:
    print(f"{country}: {count}")

# Visualization 1: Most watched genres (using listed_in as proxy)
print("\n" + "=" * 40)
print("CREATING VISUALIZATIONS")
print("=" * 40)

# Process genres
all_genres = []
for genre_list in Netflix_shows_movies['listed_in'].dropna():
    if ',' in str(genre_list):
        genres = [g.strip() for g in str(genre_list).split(',')]
        all_genres.extend(genres)
    else:
        all_genres.append(str(genre_list).strip())

genre_counter = Counter(all_genres)
top_genres = dict(genre_counter.most_common(15))

# Create genre plot
plt.figure(figsize=(12, 8))
genres = list(top_genres.keys())
counts = list(top_genres.values())

# Horizontal bar plot
bars = plt.barh(genres, counts)
plt.title('Top 15 Most Common Genres on Netflix', fontsize=14, fontweight='bold')
plt.xlabel('Number of Titles')
plt.ylabel('Genres')

# Add numbers on bars
for i, (bar, count) in enumerate(zip(bars, counts)):
    plt.text(count + 10, i, str(count), va='center')

plt.tight_layout()
plt.savefig('charts/netflix_genres_chart.png', dpi=300, bbox_inches='tight')
plt.close()

print("Genre visualization created and saved!")

# Visualization 2: Ratings distribution
rating_data = Netflix_shows_movies['rating'].value_counts()

# Create subplot with bar chart and pie chart
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6))

# Bar chart
ax1.bar(rating_data.index, rating_data.values, color='skyblue')
ax1.set_title('Netflix Content Ratings - Bar Chart')
ax1.set_xlabel('Rating')
ax1.set_ylabel('Count')
ax1.tick_params(axis='x', rotation=45)

# Add count labels on bars
for i, (rating, count) in enumerate(zip(rating_data.index, rating_data.values)):
    ax1.text(i, count + 5, str(count), ha='center')

# Pie chart
ax2.pie(rating_data.values, labels=rating_data.index, autopct='%1.1f%%', startangle=90)
ax2.set_title('Netflix Content Ratings - Distribution')

plt.tight_layout()
plt.savefig('charts/netflix_ratings_chart.png', dpi=300, bbox_inches='tight')
plt.close()

print("Ratings visualization created and saved!")

# Additional visualization - Content over time
plt.figure(figsize=(12, 6))
yearly_data = Netflix_shows_movies.groupby(['release_year', 'type']).size().unstack(fill_value=0)

# Filter for recent years (2000 onwards)
recent_data = yearly_data.loc[yearly_data.index >= 2000]

plt.stackplot(recent_data.index, 
              recent_data.get('Movie', 0), 
              recent_data.get('TV Show', 0),
              labels=['Movies', 'TV Shows'], 
              alpha=0.7)

plt.title('Netflix Content Production Over Time (2000-Present)')
plt.xlabel('Release Year')
plt.ylabel('Number of Titles')
plt.legend()
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig('charts/netflix_timeline.png', dpi=300, bbox_inches='tight')
plt.close()

print("Timeline visualization created!")

# Summary statistics
print("\n" + "=" * 40)
print("ANALYSIS SUMMARY")
print("=" * 40)

print(f"Dataset processed: {len(Netflix_shows_movies)} titles")
print(f"Movies: {len(Netflix_shows_movies[Netflix_shows_movies['type'] == 'Movie'])}")
print(f"TV Shows: {len(Netflix_shows_movies[Netflix_shows_movies['type'] == 'TV Show'])}")

print(f"\nTop 5 genres:")
for i, (genre, count) in enumerate(list(top_genres.items())[:5], 1):
    print(f"{i}. {genre}: {count}")

print(f"\nMost common rating: {rating_data.index[0]} ({rating_data.iloc[0]} titles)")

# Save cleaned dataset
Netflix_shows_movies.to_csv('data/Netflix_shows_movies_cleaned.csv', index=False)
print(f"\nCleaned dataset saved as 'data/Netflix_shows_movies_cleaned.csv'")
print("Analysis complete!")
